File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        

        stack = []

        for value in asteroids:

            while stack and stack[-1] > 0 and value < 0:

                if abs(stack[-1]) > abs(value):
                    break
                elif abs(stack[-1]) == abs(value):
                    stack.pop()
                    break
                else:
                    stack.pop()
            else:
                stack.append(value)

        return stack              




        

File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) 
    {
        int n = asteroids.size();
        std::stack<int> stack;

        for (int i = 0; i < n; i++)
        {
            bool alive = true;

            //Greater DIRECTION
            if (asteroids[i] > 0)
            {
                stack.push(asteroids[i]);
            }
            else 
            {
                //Less DIRECTION
                while (!stack.empty() && stack.top() > 0 && abs(asteroids[i]) >= abs(stack.top()))
                {
                    //SAME DESTROY BOTH
                    if (abs(asteroids[i]) == abs(stack.top()))
                    {
                        stack.pop();
                        alive = false;
                        break;
                    }

                    stack.pop();
                }

                if (!stack.empty() && stack.top() > 0 && abs(asteroids[i]) < abs(stack.top()))
                {
                    alive = false;
                }

                //Finally add
                if (alive)
                {
                    stack.push(asteroids[i]);
                }
            }
        }

        while (!stack.empty())
        {
            res.push_back(stack.top());
            stack.pop();
        }

        reverse(res.begin(), res.end());

        return res;
    }

private:
    vector<int> res;
};